### PR TITLE
Cache the getWorkerInfoList results and refresh periodically

### PR DESCRIPTION
**Pull request by LuQQiu**
getWorkerInfoList() is fast by itself, but it requires all the live worker locks and may introduce contention. 
Our clients and job servers need to use getWorkerinfoList() to select job executor, find local hostnames or know which worker to read from/write to. A large amount of `getWorkerInfoList` call may be triggered. Cache the worker info list and refresh periodically can help prevent those heavy load and lock contention

### CODE REVIEW

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 1]
```diff
@@ -2079,6 +2079,14 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.MASTER)
               .build();
+  public static final PropertyKey MASTER_WORKER_INFO_CACHE_REFRESH_TIME =
+      new Builder(Name.MASTER_WORKER_INFO_CACHE_REFRESH_TIME)
+          .setDefaultValue("10sec")
+          .setDescription("The worker information list will be refreshed "
+              + "after being cached for this time period. ")
```
Comment (gpang) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java]:
```suggestion
              + "after being cached for this time period.")
```

Also, can you add what the implications may be if it is too short, and too long?

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 2]
```diff
@@ -407,6 +431,14 @@ public long getUsedBytes() {
     if (mSafeModeManager.isInSafeMode()) {
       throw new UnavailableException(ExceptionMessage.MASTER_IN_SAFEMODE.getMessage());
     }
+    try {
+      return mWorkerInfoCache.get(WORKER_INFO_CACHE_KEY);
+    } catch (ExecutionException e) {
+      throw new UnavailableException("Unable to get worker info list from cache");
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java]: Can we include the information from `e` in this exception message?

**Comment by LuQQiu** [comment]
Jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java (hunk 2 above)
Comment (LuQQiu) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java]: Yeah, added the exception message

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (LuQQiu) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java]: DONE

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by LuQQiu** [comment]
@gpang Addressed the comments, PTAL, thanks

File: core/common/src/test/java/alluxio/ConfigurationTestUtils.java [hunk 3]
```diff
@@ -140,6 +140,9 @@ public static InstancedConfiguration defaults() {
     conf.put(PropertyKey.MASTER_PERSISTENCE_INITIAL_INTERVAL_MS, "20ms");
     conf.put(PropertyKey.MASTER_PERSISTENCE_SCHEDULER_INTERVAL_MS, "20ms");
 
+    // faster refresh
+    conf.put(PropertyKey.MASTER_WORKER_INFO_CACHE_REFRESH_TIME, "100ms");
```
Comment (gpang) [inline core/common/src/test/java/alluxio/ConfigurationTestUtils.java]: should all integration tests just use a shorter refresh time, like 20ms?

File: tests/src/test/java/alluxio/client/fs/FileOutStreamAsyncWriteIntegrationTest.java [hunk 4]
```diff
@@ -174,7 +175,8 @@ public void asyncWriteNoEvictBeforeBlockCommit() throws Exception {
       PropertyKey.Name.USER_FILE_PERSISTENCE_INITIAL_WAIT_TIME, "1min",
       PropertyKey.Name.WORKER_MEMORY_SIZE, TINY_WORKER_MEM,
       PropertyKey.Name.USER_BLOCK_SIZE_BYTES_DEFAULT, TINY_BLOCK_SIZE,
-      PropertyKey.Name.USER_FILE_BUFFER_BYTES, "8k"
+      PropertyKey.Name.USER_FILE_BUFFER_BYTES, "8k",
+      PropertyKey.Name.MASTER_WORKER_INFO_CACHE_REFRESH_TIME, "20ms"
```
Comment (gpang) [inline tests/src/test/java/alluxio/client/fs/FileOutStreamAsyncWriteIntegrationTest.java]: if the test config has a faster time, this doesn't have to be set, right?

File: core/common/src/test/java/alluxio/ConfigurationTestUtils.java (hunk 3 above)
Comment (LuQQiu) [reply core/common/src/test/java/alluxio/ConfigurationTestUtils.java]: I am worried if the too frequent getWorkerInfoList calls will create locking contentions in integration tests.

**Comment by LuQQiu** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
`alluxio.client.file.MetadataCachingBaseFileSystemTest` is failling?

**Comment by LuQQiu** [comment]
alluxio.client.file.MetadataCachingBaseFileSystemTest This test seems flaky. I don't see any relationship between this test and my changes. I tested it locally, it passed

**Comment by LuQQiu** [comment]
Jenkins, test this please

**Comment by LuQQiu** [comment]
Tests in error: 
  FileSystemUtilsIntegrationTest.alluxio.client.fs.FileSystemUtilsIntegrationTest » Runtime

**Comment by LuQQiu** [comment]
Jenkins, test this please

**Comment by LuQQiu** [comment]
Tests in error: 
alluxio.client.fs.FreeAndDeleteIntegrationTest.deleteDir(alluxio.client.fs.FreeAndDeleteIntegrationTest)
  Run 1: FreeAndDeleteIntegrationTest.deleteDir:145 » TestTimedOut test timed out after...
  Run 2: FreeAndDeleteIntegrationTest.deleteDir:145 » TestTimedOut test timed out after...

**Comment by LuQQiu** [comment]
Jenkins, test this please

**Comment by LuQQiu** [comment]
@gpang  I tested those failed tests locally, those tests all passed most of the time and occasionally they will fail. I wondering it may because our tests are flaky and also the `conf.put(PropertyKey.MASTER_WORKER_INFO_CACHE_REFRESH_TIME, "20ms");` refresh time is too small that add a lot of lock overhead.

**Comment by gpang** [comment]
why would the lock overhead matter? without this change, it would lock for every call anyways?

**Comment by LuQQiu** [comment]
That's true, it will only be refresh when the new call comes in.

**Comment by LuQQiu** [comment]
Tests in error: 
alluxio.server.ft.journal.ufs.UfsJournalIntegrationTest.directory(alluxio.server.ft.journal.ufs.UfsJournalIntegrationTest)
  Run 1: UfsJournalIntegrationTest.directory:405->directoryTestUtil:409->createFsMasterFromJournal:603 » Runtime
  Run 2: UfsJournalIntegrationTest.directory:405->directoryTestUtil:409->createFsMasterFromJournal:603 » Runtime

**Comment by LuQQiu** [comment]
@gpang Changed back to 20ms in tests and merge master to solve the flaky tests issue. PTAL, thanks

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

**Comment by LuQQiu** [comment]
alluxio-bot, cherry-pick this to branch-2.2 please

**Pull request merged**